staging/src/k8s.io/client-go/tools/cache/mutation_detector.go

```diff
@@ -113,8 +113,12 @@ func (d *defaultCacheMutationDetector) CompareObjects() {
 
 	altered := false
 	for i, obj := range d.cachedObjs {
-		if !reflect.DeepEqual(obj.cached, obj.copied) {
-			fmt.Printf("CACHE %s[%d] ALTERED!\n%v\n", d.name, i, diff.ObjectGoPrintSideBySide(obj.cached, obj.copied))
+		// make sure that we have a stable view of the cached object
+		// sometimes code will modify an object and then reset it back to the original state
+		// this can result in reflect.DeepEqual returning false while cmp.Diff prints an empty diff
+		cached := obj.cached.DeepCopyObject()
+		if !reflect.DeepEqual(cached, obj.copied) {
+			fmt.Printf("CACHE %s[%d] ALTERED!\n%v\n", d.name, i, cmp.Diff(obj.copied, cached))
```

Comment:
`cmp.Diff` won't work against objects with private fields. I know we have them in URLs in JSONSchemaProps for CRDs, not sure anywhere else. `cmp.Diff` will panic.

Reply:
we might want to catch panics here and do JSON string diff instead in those cases.

Reply:
We have API objects (runtime.Object) that are serialized on the wire that have private fields?